### PR TITLE
Closes #15 adds CDC growth chart data created by code

### LINKED ISSUE #15: General Issue: Recreate CDC growth-chart data using code
### Background Information

From discussion #13:

```
If anyone wants to re-create the reference files, the reference data for the CDC charts are at
https://www.cdc.gov/growthcharts/percentile_data_files.htm
You'd only need the data in numbers 6 (wt), ht (7), and 8 (BMI). You only need that data for sex, age, L, M, and S. But having some of the calculated percentiles might be a good check on the calculations.
[L M S are the 3 smoothing parameters to transform the data to approximate a normal distribution - they're the L (Box-Cox transformation for normality), M (median), and S (coefficient of variation)
```

Recreate the data using `tribble()` for replicability as data source may change overtime and its easier to track than floating csv's
Use snapshots for testthat 

### Definition of Done

- [ ] Render each necessary CDC dataset as specified above into an .rda object using `tibble::tribble()`
- [ ] Use `testthat::expect_snapshot()` so we can track changes in the feature

### PR DESCRIPTION
**Pull request by zdz2101**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiralpeds (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiralpeds site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the Reference page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: NEWS.md
Comment (rossfarrugia) [inline NEWS.md:1]: i think we can skip NEWS updates until we make the first release, take a look at the bottom of NEWS for admiralonco: https://github.com/pharmaverse/admiralonco/blob/main/NEWS.md

File: data-raw/cdc_metadata_creation.R
Comment (rossfarrugia) [inline data-raw/cdc_metadata_creation.R:1]: The way we did this in admiralroche was to create these as user facing exported functions and have the user run them as part of the template to create their own metadata and then feed this into their ADaM derivations. this way we don't need to store the datasets as part of the package. what's your thoughts on this @zdz2101 and @Pierre-Wallet? 
also tagging @millerg23 who leads admiralroche in case he can add any more wisdom to help us here.

Comment (millerg23) [reply data-raw/cdc_metadata_creation.R:1]: yes thats correct, users could even select what data they wanted to create.
Something like,
 `cdc_bmiage <- get_cdc_data(type = "bmiage")`
 where function get_cdc_data holds the data (created by same `tribble` statements you already have)

File: R/get_cdc_data.R
Comment (zdz2101) [inline R/get_cdc_data.R:36]: @rossfarrugia @millerg23 the same script that was used to generate the metadata is still being used but I added an argument to make it internal data, using a R/sysdata.rda file, therefore no longer exports, per r-packages textbook:

image

Comment (zdz2101) [reply R/get_cdc_data.R:36]: @Pierre-Wallet if you wanna follow along

Comment (rossfarrugia) [reply R/get_cdc_data.R:36]: @zdz2101 so you're saying our functions will be able to read in these rda metadata files internally from the package without these datasets being exported to package users right? and package users would be able to generate them themselves using the metadata functions provided or replace with their own versions? sorry Monday brain so just want to be sure i'm grasping this.

Comment (rossfarrugia) [reply R/get_cdc_data.R:36]: also let me know if you want me to take a re-review of these. happy to do some testing tomorrow

Comment (zdz2101) [reply R/get_cdc_data.R:36]: @rossfarrugia exactly what you said!

File: R/get_cdc_data.R
Comment (rossfarrugia) [inline R/get_cdc_data.R:7]: What about showing this as:
```
#'   *Permitted Values*: `"height"`, `"weight"`, `"bmi"`
```
I find it easier for users to interpret their options this way.

File: R/data.R
Comment (rossfarrugia) [inline R/data.R:1]: is this file still needed?

File: R/get_cdc_data.R
Comment (rossfarrugia) [inline R/get_cdc_data.R:9]: as a .rda? should we clarify here?

File: R/get_cdc_data.R
Comment (rossfarrugia) [inline R/get_cdc_data.R:22]:
```suggestion
#' from [CDC website](https://www.cdc.gov/growthcharts/percentile_data_files.htm)
```

File: R/get_cdc_data.R
Comment (rossfarrugia) [inline R/get_cdc_data.R:30]: Don't we usually still use `%>%` for admiral pipe convention?

File: R/get_cdc_data.R
Comment (rossfarrugia) [inline R/get_cdc_data.R:24]: we need to update man/roxygen/meta.R too for this

File: tests/testthat/_snaps/cdc_growth_charts.md
Comment (rossfarrugia) [inline tests/testthat/_snaps/cdc_growth_charts.md:1]: this unit testing approach is new to me sorry. could you explain a little how it works?

File: vignettes/admiralpeds.Rmd
Comment (rossfarrugia) [inline vignettes/admiralpeds.Rmd:1]: Instead of deleting this one (as we are going to need a Get Started guide) could you just copy over anything relevant from https://github.com/pharmaverse/admiralonco/blob/main/vignettes/admiralonco.Rmd as a starter please here? we'll need to update later once we know exactly which vignettes we'll have and any other pediatrics specifics

File: data-raw/cdc_metadata_creation.R
Comment (rossfarrugia) [reply data-raw/cdc_metadata_creation.R:1]: @zdz2101 we seem to have lost the tribble statements now which were the core part of allowing the user to recreate the metadata? am i missing something?

**Review by rossfarrugia — CHANGES_REQUESTED** [review]
@zdz2101 some initial comments through code review but will take a deeper look then from a user perspective later trying out the functions and rebuilding the pkgdown site.

File: R/data.R
Comment (zdz2101) [reply R/data.R:1]: I suppose not for this PR, https://github.com/pharmaverse/admiralpeds/pull/23 will use it, I'll delete on my PR

File: R/get_cdc_data.R
Comment (zdz2101) [reply R/get_cdc_data.R:30]: With R 4.4 coming in a month, our support going back 3 major versions would include 4.1 where the base pipe was introduced, I figured if we ever wanted to adopt native pipe, its easier to start now than swap later, I can switch it though

File: vignettes/admiralpeds.Rmd
Comment (zdz2101) [reply vignettes/admiralpeds.Rmd:1]: Added one back

File: R/get_cdc_data.R
Comment (zdz2101) [inline R/get_cdc_data.R:39]: @rossfarrugia the tribble statements are basically wrapped into here, say the user calls:

`get_cdc_data(type = "bmi")` 

the switch function will be parsed until line 39, where the return object is `cdc_bmiage` which is an internally stored dataset, which was created with the `cdc_metadata_creation.R`

File: R/get_cdc_data.R
Comment (rossfarrugia) [reply R/get_cdc_data.R:30]: there was a discussion in admiral repo on this before that i saw and it was deemed not worth to change yet as there were some cases Stefan pointed out where `%>%` works better. i forget the details. please do update

File: R/get_cdc_data.R
Comment (rossfarrugia) [reply R/get_cdc_data.R:24]: thanks @zdz2101 - we'll have to remember to come back to this later and add other categories once we add more functions

Comment (rossfarrugia) [reply R/get_cdc_data.R:24]: i wonder here if we should be more generic such as `metadata = "Metadata"`?  like admiral has it, as what if we end up adding more than just CDC/WHO growth charts here over time.

File: man/get_cdc_data.Rd
Comment (rossfarrugia) [inline man/get_cdc_data.Rd:12]: need to re-run `devtools::document()`

Comment (rossfarrugia) [reply man/get_cdc_data.Rd:12]: Also not sure this documentation is looking so clean when rendered - see below. i think we could remove the `double` references and add more to the bottom line such as `This metadata is sourced from [CDC website](https://www.cdc.gov/growthcharts/percentile_data_files.htm).`

image

File: data-raw/cdc_metadata_creation.R
Comment (rossfarrugia) [inline data-raw/cdc_metadata_creation.R:3]: I noticed something odd with the BMI only sheet from CDC where the cells are truncated in the xls for the M column only to be 8 decimal places but when you click on the values themselves they're actually given to 9 decimals, e.g. this one should be `16.575027675`. The weight and height ones are fine and all other columns look good except for this one. Please update so we have an exact match to the values from the CDC spreadsheets

**Comment by rossfarrugia** [comment]
@zdz2101 Reminder to resolve merge conflict for NEWS file. No further comments from me then besides the few remaining above.
@Pierre-Wallet please take a look too and let us know if you spot anything, or if you’re happy for me to approve/merge then once my comments are resolved. And please you can match the same conventions in your MR for the WHO charts. You can add me and Zelos as reviewers for that one then once ready.

File: data-raw/cdc_metadata_creation.R
Comment (zdz2101) [reply data-raw/cdc_metadata_creation.R:3]: ahhh the trick is that there's a difference between the .xls and .csv when you download it

Comment (rossfarrugia) [reply data-raw/cdc_metadata_creation.R]: wow, i'd guess 99.99% of the time this would never even get spotted then 😄

Comment (rossfarrugia) [reply data-raw/cdc_metadata_creation.R]: @zdz2101 you're going to hate me 😞 but i just spotted the S column has exactly the same issue! sorry, i thought it was only M!

Comment (zdz2101) [reply data-raw/cdc_metadata_creation.R]: @rossfarrugia It's been updated!

**Review by rossfarrugia — APPROVED** [review]
@zdz2101 thanks for the patience and persistence with this one! @Pierre-Wallet FYI that i'm now approving and merging this one 👍

**Comment by Pierre-Wallet** [comment]
@zdz2101 , @rossfarrugia 
sorry for my late answer, I am fine with the approach.

The only thing that is missing is the documentation (.Rd files) for each metadataset. I think it is a must have.

And one thing for my curiosity:
I tested your get_cdc_data() function and it works well but how is it possible without any call to data() function ?

File: R/get_cdc_data.R
Comment (zdz2101) [reply R/get_cdc_data.R:40]: > @rossfarrugia the tribble statements are basically wrapped into here, say the user calls:
> 
> `get_cdc_data(type = "bmi")`
> 
> the switch function will be parsed until line 39, where the return object is `cdc_bmiage` which is an internally stored dataset, which was created with the `cdc_metadata_creation.R`

@Pierre-Wallet

**Comment by zdz2101** [comment]
@Pierre-Wallet You can write the details in a generalized manner for the `.Rd` of `get_who_data()`/roclet there. 

You can store internal datasets to be retrieved in a package see this [snippet](https://r-pkgs.org/data.html#sec-data-sysdata) of the r-packages, I tagged you in a comment above that explains the logic

**Comment by Pierre-Wallet** [comment]
> @Pierre-Wallet You can write the details in a generalized manner for the `.Rd` of `get_who_data()`/roclet there.
> 
> You can store internal datasets to be retrieved in a package see this [snippet](https://r-pkgs.org/data.html#sec-data-sysdata) of the r-packages, I tagged you in a comment above that explains the logic

Yes I know how to store internal datasets, this is what I had initially proposed in my PR... The only difference is the usage of internal = TRUE parameter so that there is only one .rda instead of having one for each metadataset.

Regarding the documentation part, I would not have a generic .Rd file for the WHO metadatasets. Please not that there are 12 WHO metadatasets and they do not have the same structure. It would induce confusion for the end user.

Regarding the get_cdc_data() / get_who_data() functions, what is the actual added-value? It is only a wrapper for the data() function. Plus it brings the constraints of creating in the WHO case 12 aliases, one for each metadataset.
I understood that these functions, in the {admiralroche} case, are necessary because they contain the steps to rebuild the metadatasets within the final program, but this is not our case here. The metadatasets will remain stored anyway.
So I think we brought complexity only to handle aliases. With a proper documentation and a classical call to data(), it can be handled more easily acc to me.
Or is there another reason that I missed?

Regarding the technical point of get_cdc_data(), what I do not understand is why the dataset is loaded in the environment without any call to data() within your switch function. At first glance, I was expecting only the character string of the dataset name was returned. Do you know why?

Finally, sorry for commenting on this now, I should have jumped in before but I missed your notifications. But there are 12 WHO metadatasets to handle, so before updating everything I want to be sure on what we decide.

Thanks,
Pierre

@rossfarrugia  @zdz2101

**Comment by zdz2101** [comment]
@rossfarrugia I think it comes down to whether or not we want to export the data/how we expect users to access the datasets as well, figured if there was pre-existing way of handling metadata, we would follow suit. But I don't work with/am not familiar with {admiralroche} 

As for the technical point @Pierre-Wallet I believe the setting `LazyData: true` in the `DESCRIPTION` file is what allows this to happen

**Comment by rossfarrugia** [comment]
@zdz2101 @Pierre-Wallet my original thinking was that the 2 functions would just be tribbles and we'd never need to store the .rda in the package (that was our admiralroche approach and it worked simple enough), and then users could either call these functions or replace them with their own specific metadata created in the same format. I think in one of the discussion threads above it was felt that having the .rda as non-exported would be easier for other package components so no need to re-make them internally by running these functions. I don't feel like users should ever use the .rda files though - they should run the functions that recreate the metadatasets as dataframes. Happy if you guys both think we could simplify this way? But I might be missing something from all the discussions above and I trust both your R package dev knowledge more than mine!

**Comment by Pierre-Wallet** [comment]
> @rossfarrugia I think it comes down to whether or not we want to export the data/how we expect users to access the datasets as well, figured if there was pre-existing way of handling metadata, we would follow suit. But I don't work with/am not familiar with {admiralroche}
> 
> As for the technical point @Pierre-Wallet I believe the setting `LazyData: true` in the `DESCRIPTION` file is what allows this to happen, but I'm not 100% sure the absolute technical reason, I just thought this is how the `sysdata.rda` object works

Thanks Zelos,  I think you're right!

**Comment by Pierre-Wallet** [comment]
@rossfarrugia  thanks for your answer.

There are several options.

1. We do not store anything from CDC or WHO websites. We create functions that imports the metadatasets from their respective source.
The end users will have to run these functions all the time. It can be cumbersome and takes time because some data sets are long. It induces dependencies on the websites, which is not a good idea acc to me. We should reduce dependencies as much as possible.

1. We store the metadatasets in their original format (.csv, .xls or other) in the package. They will be in data-raw folder where the end user will not have access (because data-raw folder is not installed when end user installs the package). Some R scripts will also be in data-raw folder to read these data in their original format and create other datasets. Usually this is done to create .rda files, which will be stored in data folder. Data folder is accessible to the end user. This is the approach we have followed so far except that we did not store the metadatasets in their original formats. Instead we recreated them manually with tribble::tibble().

1. A third option is to store the metadatasets in their original format in inst folder. In this case, the end user will have access to them using system.file("metadataset.csv", package = "admiralpeds") and we can build function to import these source files into R metadatasets.

How to decide?
Scenario 2 and 3 imply a manual operation when recreating the original metadata. This is a source of error. And the community may not trust us. Because it relies entirely on Zelos and I :)
There could be an option to guarantee that the recreated metadatasets stick to their website source by using a test. I have never done that but I think it is possible to test that our internal metadatasets are identical to the imported ones from the website. This plus the snapshot test should be enough to convince the community that we are reliable.

I do not like scenario 3 as there is a small added-value of keeping the metadatasets in their original formats as the end user can access them easily on the respective websites.
I am not in favor of scenario 1 because we do not leverage all the capacities of a package. Plus having the data available in .rda format makes everything easier and more flexible, both for the developers and the end users.
I am in favor of scenario 2, assuming the tests I discussed above are feasible. But in that case, we do not need the get_cdc_data() / get_who_data() because they do not bring any added value.

I hope I am clear enough.

Do not hesitate to set up a quick meeting with the 3 of us if you feel it is needed.
Happy to elaborate.

**Comment by rossfarrugia** [comment]
@zdz2101 as you're technical lead though I trust your call as to whether you prefer to swap out `get_cdc_data()` and instead give the users these files as .rda.  I know in admiral package there are some .rda files provided to users. The main point though is let's be transparent with our users how these were generated so that they have access to whatever was used and they could adapt it if ever needed to supply their own metadata.

**Comment by millerg23** [comment]
Hi All,
Interesting conversation, I think option 2 is preferable, I like the idea of storing the original data in the package.
We did something similar in `admiral` core package for lab grading, where the spreadsheets from NCI were added to the package, with some extra columns added by us. We then read in the spreadsheets and stored them as .rda files, and are accessed directly in the code, we don't even need `data()` call.
I also lead the admiralroche package, where we had several (smaller) pieces of metadata and these were created by `tibble` function, and it made sense to have `get_<metadata>` functions, for the reasons Ross described. We also didn't have "original" spreadsheets for these metadata, as they are internal to Roche, and users may want to overwrite some of them.
If you remove the `get_` functions, as long as the `.rda` files are well documented and users can easily see what they are called, I think this is fine also. 

Regards,
Gordon.

**Comment by Pierre-Wallet** [comment]
@rossfarrugia @zdz2101 @millerg23 
Hi everyone,

Sorry Ross but I do not understand what you expect.
You mentioned you're in favor of 2, which means having the metadata in .rda but at the end you would vote not to have the metadata in .rda. This is contradictory.

If I understood correctly the get_<metadata> behaviour, you expect this function to recreate the metadata with tribble::tibble(). This is not what we have done so far. The get_cdc_data() is just loading the metadata already existing in data folder. 
Please correct me if I misunderstood.

Regarding the testing part, sorry but I did not follow all the discussions you had on Zelos PR. 
I totally agree with the electronic comparison you performed. I just mentioned it would be good to automatize it in the unit tests in testthat folder.

Now if you want to review my PR, feel free to do it right now because it already reflects the scenario I am in favor:
- the tribble::tibble() statements are available in data_raw folder, they create the metadata in data folder
- the documentation for each metadaset is available
The only missing elements  are head circumference data, get everything in one .rda file and the automatization of the test so that you do not even have to do it manually in your PR review.

Kind Regards,
Pierre.

**Comment by Pierre-Wallet** [comment]
@zelos, available on Monday afternoon for me, I think it will be morning for you assuming you're US based.

@rossfarrugia thanks for the clarifications. At least, I have confirmation that I understood correctly your needs.

We'll make it.

**Pull request merged**
